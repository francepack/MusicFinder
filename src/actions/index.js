export const storeBand = (band) => ({
  type: 'STORE_BAND',
  band
})

export const storeSimilarBands = (bands) => ({
  type: 'STORE_SIMILAR',
  bands
})

export const storeBandTags = (tags) => ({
  type: 'STORE_TAGS',
  tags
})

export const storeEvents = (events) => ({
  type: 'STORE_EVENTS',
  events
})