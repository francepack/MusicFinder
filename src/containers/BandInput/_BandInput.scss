form {
  margin-top: 60px;
}

.band-input {
  @include inputs;
  border-radius: 10px 0 0 10px; 
  width: 40%;
}

.submit-btn {
  @include btn
  border-radius: 0 10px 10px 0;
  width: 80px;
}