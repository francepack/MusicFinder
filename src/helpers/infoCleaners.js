import React from 'react'

export const createUrlString = (keyword, state, city, startDate, endDate) => {
  const keywordUrl = makeStringUrl(keyword)
  let urlString = `&classificationName=music&keyword=${keywordUrl}`
  if (state) {
    const cleanState = makeStringUrl(state)
    const stateUrl = `&stateCode=${cleanState}`
    urlString = urlString + stateUrl
  }
  if (city) {
    const cleanCity = makeStringUrl(city)
    const cityUrl = `&city=${cleanCity}`
    urlString = urlString + cityUrl
  }
  if (startDate && startDate.split('/').length === 3) {
    const cleanStartDate = makeDateUrl(startDate)
    const startDateUrl = `&startDateTime=${cleanStartDate}T01:00:00Z`
    urlString = urlString + startDateUrl
  }
  if (endDate && endDate.split('/').length === 3) {
    const cleanEndDate = makeDateUrl(endDate)
    const endDateUrl = `&endDateTime=${cleanEndDate}T01:00:00Z`
    urlString = urlString + endDateUrl
  }
  return urlString
}

export const makeStringUrl = (band) => {
  const letters = band.split('')
  const url = letters.map(letter => {
    switch (letter) {
      case ' ':
        return '+'
      case '!':
        return '%21'
      case '"':
        return '%22'
      case '#':
        return '%23'
      case '$':
        return '%24'
      case '%':
        return '%25'
      case '&':
        return '%26'
      case "'":
        return '%27'
      case '(':
        return '%28'
      case ')':
        return '%29'
      case '*':
        return '%2A'
      case '+':
        return '%2B'
      case '/':
        return '%2F'
      default:
        return letter
    }
  })
  return url.join('')
}

export const makeDateUrl = (date) => {
  const splitDate = date.split('/')
    const year = splitDate[2]
    const day = splitDate[1]
    const month = splitDate[0]
    return `${year}-${month}-${day}`
}

export const matchSimilarBands = (lastfmArr, tastediveArr) => {
  if ( lastfmArr.length && tastediveArr.length) {
    return lastfmArr.reduce((acc, val) => {
      let matchedBand = tastediveArr.find(band => {
        return band === val
      })
      if (matchedBand) acc.push(matchedBand)
      return acc
    }, [])
  } else if (!lastfmArr.length && tastediveArr.length) {
    return tastediveArr.slice(0, 10)
  } else if (!tastediveArr.length && lastfmArr.length) {
    return lastfmArr.slice(0, 10)
  }
}

export const buildBandArray = (matchedBands, tastebands, lastfmbands) => {
  const arrayLength = matchedBands.length
  if (arrayLength === 10) {
    return matchedBands
  } 
  
  else if (arrayLength === 0) {
    if (tastebands.length) {
      return tastebands.slice(0, 10)
    } else if (!tastebands.length && lastfmbands.length) {
      return lastfmbands.slice(0, 10)
    } else if (!tastebands.length && !lastfmbands.length) {
      return []
    }
  } 
  
  else if (arrayLength > 10) {
    return matchedBands.slice(0, 10)
  } 
  
  else if (arrayLength < 10) {
    let count = 0
    for (let i = arrayLength; i < 10; i++) {
      count++
    }
    if (tastebands.length) {
      let filteredBands = tastebands.reduce((acc, band) => {
        if (acc.length < count && !matchedBands.includes(band)) {
          acc.push(band)
        }
        return acc
      }, [])
      return matchedBands.concat(filteredBands)
    } else if (!tastebands.length && lastfmbands.length) {
      let filteredBands = lastfmbands.reduce((acc, band) => {
        if (acc.length < count && !matchedBands.includes(band)) {
          acc.push(band)
        }
        return acc
      }, [])
      return matchedBands.concat(filteredBands)
    } else if (!tastebands.length && !lastfmbands.length) {
      return matchedBands
    }
  }
}

export const cleanEvents = (event) => {
  let venueName = 'No venue listed'
  let venueCity = 'No city listed'
  let venueAddress = 'No address given'
  let date = 'No date listed'
  let eventImg
  if (event._embedded.venues) {
    venueName = event._embedded.venues[0].name
    venueCity = event._embedded.venues[0].city.name
    venueAddress = event._embedded.venues[0].address.line1
  }
  if (event.images) {
    eventImg = event.images[0].url
  } else {
    eventImg = 'https://image.freepik.com/free-vector/blue-background-people-concert_23-2147604883.jpg'
  }
  if (event.dates.start.localDate) {
    let splitDate = event.dates.start.localDate.split('-')
    date = `${splitDate[1]}/${splitDate[2]}/${splitDate[0]}`
  }

  return { 
    name: event.name, 
    eventUrl: event.url, 
    id: event.id, 
    date: date,
    venue: venueName,
    venueAddress: venueAddress,
    city: venueCity,
    image: eventImg
  }
  
}

export const buildCards = (events) => {
  if (events.length === 1) {
    const event = events[0]
    const background = { backgroundImage: `url(${event.image})` }
    return(
      <main className='single'>
        <section className='single-event-card' key={event.id}>
          <a href={event.eventUrl} target='_blank' rel='noopener noreferrer'>
            <div className='background' style={background}>
              <div className='overlay'>
                <article className='card-details'>
                  <header className='top-details'>
                    <h3>{event.name}</h3>
                    <p>{event.date}</p>
                  </header>
                  <footer className='bottom-details'>
                    <p>City: {event.city}</p>
                    <p>Venue: {event.venue}</p>
                  </footer>
                </article>
              </div>
            </div>
          </a>
        </section>
      </main>
    )
  }
  const cards = events.map(event => {
    const background = { backgroundImage: `url(${event.image})` }
    return(
      <section className='event-card' key={event.id}>
        <a href={event.eventUrl} target='_blank' rel='noopener noreferrer'>
          <div className='background' style={background}>
            <div className='overlay'>
              <article className='card-details'>
                <header className='top-details'>
                  <h3>{event.name}</h3>
                  <p>{event.date}</p>
                </header>
                <footer className='bottom-details'>
                  <p>City: {event.city}</p>
                  <p>Venue: {event.venue}</p>
                </footer>
              </article>
            </div>
          </div>
        </a>
      </section>
    )
  })
  return(<main className='events-container'>{cards}</main>)
}