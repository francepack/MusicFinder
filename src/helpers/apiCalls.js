import { lastfmKey, eventfulKey, ticketmasterKey, tastediveKey } from '../apiKey/key'
import { makeBandUrl, buildBandArray, cleanEvents } from './infoCleaners'

export const getSimilarBands = async (band) => {
  let bandUrl = makeBandUrl(band)
  try {
    let tastediveBands = await tastediveGetSimilarBands(bandUrl)
    let lastfmBands = await lastfmGetSimilarBands(bandUrl)
    let similarBands = lastfmBands.reduce((acc, val) => {
      let matchedBand = tastediveBands.find(band => {
        return band === val
      })
      if (matchedBand) acc.push(matchedBand)
      return acc
    }, [])
    let bandsArray = await buildBandArray(similarBands, tastediveBands) 
    return bandsArray
  } catch(error) {
    return error.message
  }
}

const lastfmGetSimilarBands = async (bandUrl) => {
  const url = `http://ws.audioscrobbler.com/2.0/?method=artist.getsimilar&artist=${bandUrl}&api_key=${lastfmKey}&format=json`
  try {
    let response = await fetch(url)
    let similarBands = await response.json()
    let bandNames = await similarBands.similarartists.artist.map(band => band.name)
    return bandNames
  } catch(error) {
    return error.message
  }
}

const tastediveGetSimilarBands = async (bandUrl) => {
  const proxyUrl = `https://cors-anywhere.herokuapp.com/`
  const url = `https://tastedive.com/api/similar?q=${bandUrl}&type=music&k=${tastediveKey}`
  try {
    let response = await fetch(proxyUrl + url)
    let similarBands = await response.json()
    let bandNames = await similarBands.Similar.Results.map(band => band.Name)
    return bandNames
  } catch(error) {
    return error.message
  }
}

export const getBandTags = async (band) => {
  const url = `http://ws.audioscrobbler.com/2.0/?method=artist.gettoptags&artist=${band}&user=RJ&api_key=${lastfmKey}&format=json`
  try {
    let response = await fetch(url)
    let bandTags = await response.json()
    let tags = await bandTags.toptags.tag.slice(0,10).map(tag => tag.name)
    return tags
  } catch(error) {
    return error.message
  }
}

export const getEvents = async (urlString) => {
  const url = `https://app.ticketmaster.com/discovery/v1/events.json?${urlString}&apikey=${ticketmasterKey}`
  try {
    const response = await fetch(url)
    const eventData = await response.json()
    console.log(eventData)
    const events = await cleanEvents(eventData)
    console.log(events)
    // const ImageUrls = await findImageUrls(events)
  } catch(error) {
    return error.message
  }
}

// const findImageUrls = (events) => {
//   let images = events.map(async event => {
//     let url = `https://s1.ticketm.net${event.image}`
//     try {
//       console.log(event.image)
//       // await fetch(event.image)
//     } catch(error) {
//       return error.message
//     }
//   })

// }











// export const getInfo = async () => {
//   const url = `http://ws.audioscrobbler.com/2.0/?method=tag.gettopartists&tag=rock&api_key=${lastfmKey}&format=json`
//   try {
//     const response = await fetch(url)
//     const info = await response.json()
//     return info.topartists.artist.map(art => art.name)
//   } catch (error) {
//     return error.message
//   }
// }

// export const getEventfulBandId = async () => {
//   const url = `http://api.eventful.com/json/performers/search?app_key=${eventfulKey}&keywords=ween`
//   const proxyurl = "https://cors-anywhere.herokuapp.com/"
//   try {
//     const response = await fetch(proxyurl + url)
//     const info = await response.json()
//     return info.performers.performer.map(art => art.id)
//   } catch (error) {
//     return error.message
//   }
// }

// export const getDenverEvents = async () => {
//   const url = `http://api.eventful.com/json/events/search?app_key=${eventfulKey}&category=music&l=Denver`
//   const proxyurl = "https://cors-anywhere.herokuapp.com/"
//   try {
//     const response = await fetch(proxyurl + url)
//     const info = await response.json()
//     return info.events.event.map(event => event.title)
//   } catch(error) {
//     return error.message
//   }
// }


// export const ticketmasterevents = async () => {
//   const url = `https://app.ticketmaster.com/discovery/v1/events.json?apikey=${ticketmasterKey}`
//   // const proxyurl = "https://cors-anywhere.herokuapp.com/"
//   try {
//     const response = await fetch(url)
//     const info = await response.json()
//     return info._embedded.events.map(event => event.name)
//   } catch(error) {
//     return error.message
//   }
// }