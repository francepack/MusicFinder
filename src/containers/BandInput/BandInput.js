import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'
import { storeBand, storeSimilarBands, storeBandTags } from '../../actions'
import { getSimilarBands, getBandTags } from '../../helpers/apiCalls'

export class BandInput extends Component {
  constructor(props) {
    super(props)
    this.state = {
      bandInput: ''
    }
  }

  handleChange = (e) => {
    this.setState({
      [e.target.name]: e.target.value
    })
  }

  handleSubmit = (e) => {
    e.preventDefault()
    this.findResults()
  }

  findResults = () => {
    const { bandInput } = this.state
    this.props.setLoading()
    this.saveBand(bandInput)
    this.getSearchIdeas(bandInput)
  }

  saveBand = (band) => {
    this.props.storeBand(band)
  }

  getSearchIdeas = async (band) => {
    try {
      const { history } = this.props
      let similarBands = await this.searchSimilarBands(band)
      let bandTags = await this.searchForBandTags(band)
      this.props.storeSimilarBands(similarBands)
      this.props.storeBandTags(bandTags)
      history.push('/band-info')
      this.props.setLoading()
    } catch(error) {
      return error.message
    }
  }

  searchSimilarBands = async (band) => {
    try {
      const similarBands = await getSimilarBands(band)
      return similarBands
    } catch(error) {
      return error.message
    }
  }

  searchForBandTags = async (band) => {
    try {
      const bandTags = await getBandTags(band)
      return bandTags
    } catch(error) {
      return error.message
    }
  }

  render() {
    return(
      <form onSubmit={this.handleSubmit}>
        <input  className='band-input'
                name='bandInput'
                placeholder='Enter a Band'
                defaultValue={this.state.bandInput}
                onKeyUp={this.handleChange}
                >
        </input>
        <button className='submit-btn'>Submit</button>
      </form>
    )
  }
}

export const mapDispatchToProps = (dispatch) => ({
  storeBand: (band) =>  dispatch(storeBand(band)),
  storeSimilarBands: (bands) => dispatch(storeSimilarBands(bands)),
  storeBandTags: (tags) => dispatch(storeBandTags(tags))
})

BandInput.propTypes = {
  storeBandTags: PropTypes.func,
  storeSimilarBands: PropTypes.func,
  storeBand: PropTypes.func,
  setError: PropTypes.func,
  setLoading: PropTypes.func
}

export default withRouter(connect(null, mapDispatchToProps)(BandInput))