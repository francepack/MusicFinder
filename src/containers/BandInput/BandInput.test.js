import React from 'react'
import { shallow } from 'enzyme'
import { BandInput, mapDispatchToProps } from './BandInput'
import * as actions from '../../actions'
import { getSimilarBands, getBandTags } from '../../helpers/apiCalls';

jest.mock('../../helpers/apiCalls')


describe('BandInput', () => {
  let wrapper
  let props

  beforeEach(() => {
    props = {
      setLoading: jest.fn(),
      storeBand: jest.fn(),
      storeSimilarBands: jest.fn(),
      storeBandTags: jest.fn()
    }
    wrapper = shallow(
      <BandInput {...props} />
    )
  })
  describe('defaultstate', () => {
    it('should have default state', () => {
      const expected = {bandInput: ''}
      expect(wrapper.state()).toEqual(expected)
    })
  })
  describe('Snapshots', () => {
    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('saveBand', () => {
    it('should call prop storeBand', () => {
      const mockBand = 'ween'
      wrapper.instance().saveBand(mockBand)
      expect(wrapper.instance().props.storeBand).toHaveBeenCalledWith(mockBand)
    })
  })
  describe('searchSimilarBands', () => {
    it('should call getSimilarBands with a bandname', () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands(mockBand)
      expect(getSimilarBands).toHaveBeenCalledWith(mockBand)
    })
    it('should return an error message if failed', () => {
      
    })
  })
  describe('searchForBandTags', () => {
    it('should call getBandTags with a bandname', () => {
      const mockBand = 'Ween'
      wrapper.instance().searchForBandTags(mockBand)
      expect(getBandTags).toHaveBeenCalledWith(mockBand)
    })
    it('should return an error message if failed', () => {
      
    })
  })
  describe('getSearchIdeas', () => {
    it('should call searchSimilarBands', () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands = jest.fn()
      wrapper.instance().getSearchIdeas(mockBand)
      expect(wrapper.instance().searchSimilarBands).toBeCalled()
    })
    it('should call searchForBandTags', async () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands = jest.fn()
      wrapper.instance().searchForBandTags = jest.fn()
      await wrapper.instance().getSearchIdeas(mockBand)
      expect(wrapper.instance().searchForBandTags).toBeCalled()
    })
    it('should call props storeSimilarBands', async () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands = jest.fn()
      wrapper.instance().searchForBandTags = jest.fn()
      await wrapper.instance().getSearchIdeas(mockBand)
      expect(wrapper.instance().props.storeSimilarBands).toBeCalled()
    })
    it('should call props storeBandTags', async () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands = jest.fn()
      wrapper.instance().searchForBandTags = jest.fn()
      await wrapper.instance().getSearchIdeas(mockBand)
      expect(wrapper.instance().props.storeBandTags).toBeCalled()
    })
    it('should call props setLoading', async () => {
      const mockBand = 'Ween'
      wrapper.instance().searchSimilarBands = jest.fn()
      wrapper.instance().searchForBandTags = jest.fn()
      await wrapper.instance().getSearchIdeas(mockBand)
      expect(wrapper.instance().props.setLoading).toBeCalled()
    })
    it('should return an error message if failed', () => {
      
    })
  })
  describe('handleSubmit', () => {
    it('should prevent default', () => {
      let mockEvent = { preventDefault: jest.fn() }
      wrapper.instance().handleSubmit(mockEvent)
      expect(mockEvent.preventDefault).toBeCalled()
    })
    it('should call findResults', () => {
      wrapper.instance().findResults = jest.fn()
      let mockEvent = { preventDefault: jest.fn() }
      wrapper.instance().handleSubmit(mockEvent)
      expect(wrapper.instance().findResults).toBeCalled()
    })
  })
  describe('handleChange', () => {
    it('should set bandInput state as user types', () => {
      const mockEvent = {target: {name: 'bandInput', value: 'hello'}}
      expect(wrapper.state()).toEqual({bandInput: ''})
      wrapper.instance().handleChange(mockEvent)
      expect(wrapper.state()).toEqual({bandInput: 'hello'})
    })
  })
  describe('mapDispatchToProps', () => {
    it('should store a band', () => {
      const mockDispatch = jest.fn()
      const mockBand = 'Ween'
      const actionToDispatch = actions.storeBand(mockBand)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.storeBand(mockBand)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
    it('should store suggested similar bands', () => {
      const mockDispatch = jest.fn()
      const mockBands = ['Weezer', 'Cake', 'ExampleBand']
      const actionToDispatch = actions.storeSimilarBands(mockBands)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.storeSimilarBands(mockBands)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
    it('should store band descriptions', () => {
      const mockDispatch = jest.fn()
      const mockTags = ['so fresh', 'so clean']
      const actionToDispatch = actions.storeBandTags(mockTags)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.storeBandTags(mockTags)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})