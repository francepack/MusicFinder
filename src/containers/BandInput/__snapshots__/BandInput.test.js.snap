// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BandInput Snapshots should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <BandInput
    setLoading={[MockFunction]}
    storeBand={[MockFunction]}
    storeBandTags={[MockFunction]}
    storeSimilarBands={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="band-input"
          defaultValue=""
          name="bandInput"
          onKeyUp={[Function]}
          placeholder="Enter a Band"
        />,
        <button
          className="submit-btn"
        >
          Submit
        </button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "band-input",
          "defaultValue": "",
          "name": "bandInput",
          "onKeyUp": [Function],
          "placeholder": "Enter a Band",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Submit",
          "className": "submit-btn",
        },
        "ref": null,
        "rendered": "Submit",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="band-input"
            defaultValue=""
            name="bandInput"
            onKeyUp={[Function]}
            placeholder="Enter a Band"
          />,
          <button
            className="submit-btn"
          >
            Submit
          </button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "band-input",
            "defaultValue": "",
            "name": "bandInput",
            "onKeyUp": [Function],
            "placeholder": "Enter a Band",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "className": "submit-btn",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
